[JSC] Inline operationWasmIsStrictSubRTT

Constellation opened the pull request:
#### e53d8180afea3b4564723151232d48e166a72003
<pre>
[JSC] Inline operationWasmIsStrictSubRTT
<a href="https://bugs.webkit.org/show_bug.cgi?id=297482">https://bugs.webkit.org/show_bug.cgi?id=297482</a>
<a href="https://rdar.apple.com/158436646">rdar://158436646</a>

Reviewed by Keith Miller.

Inline operationWasmIsStrictSubRTT into OMG code. This is simply Cohen&apos;s
type check, and we can fully inline it with very small amount of machine
code.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallIndirect):
(JSC::Wasm::BBQJITImpl::BBQJIT::addRTTSlowPathJump): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::emitSlowPathRTTCheck):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::emitRefTestOrCast):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::RTT::RTT):
(JSC::Wasm::RTT::tryCreate):
(JSC::Wasm::RTT::isSubRTT const):
(JSC::Wasm::RTT::isStrictSubRTT const):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm:: const):

Canonical link: <a href="https://commits.webkit.org/298798@main">https://commits.webkit.org/298798@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9abd196e6e75e7fa2e8d6c365936799ae742c33f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116686 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26913 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122760 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67259 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ca9469c3-3a89-4fdb-992f-fb1b8dfaeb9b) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37048 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44939 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88619 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9fac5183-8e21-4f82-89ad-f0e47b51f139) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119635 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29543 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104674 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69087 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28605 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22780 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66427 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108800 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98923 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22935 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125896 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115214 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43585 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97283 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43949 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100877 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97077 "Found 10 new API test failures: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/open-window-default-size, /WebKitGTK/TestWebKitFindController:/webkit/WebKitFindController/hide, /TestWebKit:WebKit.OnDeviceChangeCrash, /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/snapshot, /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/preferred-size, /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/custom-container-destroyed, /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, /TestWebKit:WebKit.GeolocationTransitionToHighAccuracy, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/test-page-list, /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/page-visibility (failure)") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42403 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20329 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18631 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43471 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49066 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143912 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42938 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/37044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46277 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/44643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b2d592ed9e3c64918c6fa8665500c9232aeb2167. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49479

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7952c5a27270ec3f6553b0dd68be5a102302d556. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49479

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -4618,9 +4598,14 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addCallIndirect(unsigned tableIndex, co
+            if (needsSubtypeCheck) {
+                auto jump = m_jit.jump();
+                auto returnLabel = m_jit.label();
+                m_latePaths.append([jump, returnLabel, signature = originalSignature.index(), calleeRTT](BBQJIT& bbq, CCallHelpers& jit) {
+                    jump.link(jit);
+                    bbq.emitSlowPathRTTCheck(returnLabel, signature, calleeRTT);
+                });
```
This doesn't seem like something that should be a late path. It could happen 100% of the time.

Constellation replied:
Yeah, removed the late path.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -611,18 +611,28 @@ bool TypeDefinition::isFinalType() const
-    for (unsigned i = 0; i < supertype.displaySize(); ++i)
-        at(i + 1) = supertype.at(i);
+    ASSERT(supertype.size() == (supertype.displaySizeExcludingThis() + 1));
+    for (unsigned i = 0; i < supertype.size(); ++i)
+        at(i) = supertype.at(i);
```
Nit: This could be a memcpySpan.

kmiller68 reviewed (approved):
r=me with nits.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9abd196e6e75e7fa2e8d6c365936799ae742c33f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49479

Constellation committed:
[JSC] Inline operationWasmIsStrictSubRTT
https://bugs.webkit.org/show_bug.cgi?id=297482
rdar://158436646

Reviewed by Keith Miller.

Inline operationWasmIsStrictSubRTT into OMG code. This is simply Cohen's
type check, and we can fully inline it with very small amount of machine
code.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addCallIndirect):
(JSC::Wasm::BBQJITImpl::BBQJIT::addRTTSlowPathJump): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::emitSlowPathRTTCheck):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::emitRefTestOrCast):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::RTT::RTT):
(JSC::Wasm::RTT::tryCreate):
(JSC::Wasm::RTT::isSubRTT const):
(JSC::Wasm::RTT::isStrictSubRTT const):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm:: const):

Canonical link: https://commits.webkit.org/298798@main

webkit-commit-queue commented:
Committed 298798@main (e53d8180afea): <https://commits.webkit.org/298798@main>

Reviewed commits have been landed. Closing PR #49479 and removing active labels.
